libraries/chain/controller.cpp

```diff
@@ -2592,20 +2592,51 @@ block_state_ptr controller::finalize_block( const signer_callback_type& signer_c
 
    auto& ab = std::get<assembled_block>(my->pending->_block_stage);
 
+   auto pfa = *ab._pending_block_header_state.prev_activated_protocol_features.get();
+   auto pfs = my->protocol_features.get_protocol_feature_set();
+
    auto bsp = std::make_shared<block_state>(
                   std::move( ab._pending_block_header_state ),
                   std::move( ab._unsigned_block ),
                   std::move( ab._trx_metas ),
-                  my->protocol_features.get_protocol_feature_set(),
+                  pfs,
                   []( block_timestamp_type timestamp,
                       const flat_set<digest_type>& cur_features,
                       const vector<digest_type>& new_features )
-                  {},
-                  signer_callback
+                  {}
               );
 
    my->pending->_block_stage = completed_block{ bsp };
 
+   if( my->block_sign_fut.valid() ) {
```

Comment:
This is not happening until the next block. I don't think we want to delay emitting the signed block by 500ms. This all should happen as soon as the block is signed. Can all this be moved into the lambda passed to block_sign_pool?

Reply:
The signal emission is thread-safe but `log_irreversible()` uses `boost::multi_index_container` in the fork database which is not remotely thread-safe.  Emitting the block to the network early results in the peculiar case of the network possibly having an irreversible block which the local node does not consider irreversible for hundreds of milliseconds.  If a health monitor in a single producer network happens to shut down and restart a producing node in that window without activating a standby producer, the network will irreconcilably fork as the active producer essentially double-produces a block, having lost its own prior last block.

Reply:
I'm envisioning it working as follows:
- kick off a lambda to sign and do whatever else is possible in a diff thread
- When signing done it will post back to the main thread at high priority to finish out the `log_irreversible()` and `emit`.
- Indicate block is complete and block on the next block at that location if the previous is still signing (should not happen in practice as it should not take more than 500ms to sign, although we should setup a test that causes it to happen by having a very slow sign proces..)